### PR TITLE
NetPlay: Add support for user-triggered events.

### PR DESCRIPTION
**Pull request by AdmiralCurtiss**
This does two things (split into two commits):

- In `CoreTiming`, adds a second event type that is executed like the regular events but does not participate in the timing calculation for when `Advance()` is called next. This allows scheduling an event to run at an explicit emulated time while remaining deterministic even if only a single client runs it in a netplay scenario, as long as the event does not affect the emulated state. (This could also be used for things like scripting or GDB, by the way!)
- In NetPlay, adds logic to schedule an event and sync its execution to happen at the same emulated timepoint on all clients.

This allows users to trigger events during netplay without desyncing netplay. Currently I've only implemented a single simple event type (pressing the Reset button on the emulated console, which would previously only run on the client that clicked the menu option, immediately desyncing the whole netplay session), but I'm hoping to implement more stuff in the future, such as savestates or reassigning controller ports mid-session.

---

On the technical side, the CoreTiming change is pretty straightforward, but the NetPlay one is a bit hard to follow, so I've made a diagram to help better understand it:

Untitled

There's a timeout (of 5 seconds) in the syncing function so that if, for whatever reason, a client does not send their timepoint the emulation doesn't freeze forever waiting for the response. -- I've had this happen in testing once because one instance was far enough ahead of another so it stalled waiting for input before it got to the syncing event, so both clients were waiting for the other to send a different thing. If this happens, the event is cancelled. To make this less likely in practice I scheduled the syncing event ~1 second after the user actually initiates the event (previously I've had it scheduled immediately), so that it's more likely that each client receives the event before it's actually scheduled. The extra timepoint syncing is still required because we have no way to guarantee that the event *actually* ends up arriving before it is scheduled.

If someone has a better idea of how to accomplish this, I'm open to suggestions, but this appears to work pretty well.

### CODE REVIEW

**Comment by mimimi085181** [comment]
Is it possible for your code to cause stutter, because the client has promised to schedule the event at a certain time, and it has reached that time, so it needs to wait for the other clients?

The way i would have implemented it, if i was skilled enough, would have abused the input buffer. The inputs are already synced, so if a special input would trigger an event, that event would also be synced. That special input would need to be inside the regular input or an additional input, so the number of controller inputs is not disturbed.

One drawback would be that only players with an active input device are able to schedule events. Maybe that's exactly what's wanted, maybe not. With your idea, it would be possible for spectators to trigger events, as long as you allow them to.

One potentially big advantage would be that events like pressing the reset button would be exactly in order with the inputs. So press A, reset, B, would do exactly that. With your method it would properly sync, but it would most likely result in A, B, reset. An extreme example would be homebrew software that uses the power and reset buttons as inputs. One note here: I would add the event triggering in the gamecube and wiimote input handling, and the 1st one to see the event request to send it to the other clients.

Another alternative would be a special buffer for events, but that might either over complicate the code or double the network packets, while 99.99% of them are nops.

What about syncing shutdown? How is that done currently? Would there be an advantage of syncing shutdowns at precisely the same emulated time?

And does switching discs work over netplay now? In theory it should work if Dolphin auto-switches to the 2nd disc(i think i asked this already somewhere but forgot the answer). Action replay over netplay would be nice for some people i think.

**Comment by AdmiralCurtiss** [comment]
> Is it possible for your code to cause stutter, because the client has promised to schedule the event at a certain time, and it has reached that time, so it needs to wait for the other clients?

Yes. I don't see a way around that though; you could make the system a bit smarter and send around more data in advance to reduce the chance, but there has to be a defined sync point *somewhere* where all clients agree to do the same thing before they continue emulating past that.

> The way i would have implemented it, if i was skilled enough, would have abused the input buffer. [...]

I was also thinking about that but I came to the conclusion that that's a worse approach; most events are likely controller-agnostic and most definitely input-poll-time agnostic. Say if you got the game into a state where it wouldn't poll inputs anymore, you could also no longer send eg. a savestate load event. Additionally, my primary motivation for this was the ability to change the controller assignments at runtime, up to and including passing a single controller between multiple users, so only allowing players with active inputs to send events seemed counterproductive to that.

You're right that this approach does not guarantee a timed sequence between events and controller input -- or at least not one obvious to the user. I don't know if that's really a problem in practice though.

> What about syncing shutdown? How is that done currently? Would there be an advantage of syncing shutdowns at precisely the same emulated time?

Shutdown is currently pretty unsynced, which did lead to a pretty ugly bug with Wii safe shutdown that I only recently worked around in #11059. So yes, I think it would be good to sync that correctly.

> And does switching discs work over netplay now? In theory it should work if Dolphin auto-switches to the 2nd disc(i think i asked this already somewhere but forgot the answer). Action replay over netplay would be nice for some people i think.

I have no idea how any of this works currently.

**Comment by iwubcode** [comment]
@AdmiralCurtiss - do we have anyone that would be well suited to reviewing this?  If not, I can at least take a look from a code perspective (netplay has been something I've avoided getting into despite being pretty interested in it).

**Comment by AdmiralCurtiss** [comment]
I believe I've mentioned on Discord that I'm not entirely happy with the implementation here. It's better than nothing, definitely, so I guess we can take it, and then swap it out for something better once we do have a better option...

So yeah if someone wants to review this go ahead.
